### ISSUE TITLE
[Bug][SFT][FSDP] max_ckpt_to_keep=1 is not enforced after resume

### ISSUE THREAD

**Issue by ai-yang** [verl-project/verl#7395]
## Severity

High for preemptible or repeatedly resumed training. Each restart leaves the previously loaded model and optimizer shards on disk, so large checkpoints can accumulate until the filesystem is exhausted and a later save or training job fails.

## System Info

- Reproduced on latest verl main commit: `09ac37258ea66b0cb69b2738eec3074ea4e7261c`
- Proposed branch commit: `8a1a031a9c03105839dcc6659e15cd975bfd5246`
- Linux x86_64, kernel 5.15
- 4 × NVIDIA GeForce RTX 3090 (24 GiB), driver 580.173.02
- Python 3.12.13
- PyTorch 2.11.0+cu128, CUDA 12.8
- Ray 2.56.1
- Transformers 5.10.4
- Model: Qwen3.5-0.8B, local revision `2fc06364715b967f1860aea9cf38778875588b17`
- SFT Ray trainer, FSDP2 size 4, synchronous checkpoint save

## Information

- [x] The SFT Ray trainer
- [x] The FSDP2 checkpoint path
- [x] A deterministic CPU reproduction
- [x] A real four-GPU restart reproduction

## Reproduction

The minimal reproducer creates an existing step-1 SFT checkpoint, then uses a fresh production `BaseCheckpointManager` and `CheckpointHandler` to load it and save step 2, modeling a new resumed process:

```python
import os
import tempfile
from unittest.mock import patch

import torch
from verl.utils.checkpoint.checkpoint_handler import CheckpointHandler, OrchestrationMode
from verl.utils.checkpoint.checkpoint_manager import BaseCheckpointManager

class Dataloader:
    def state_dict(self): return {"position": 1}
    def load_state_dict(self, state): assert state == {"position": 1}

class Engine:
    def __init__(self, manager): self.manager = manager
    def load_checkpoint(self, local_path, **kwargs): assert os.path.isdir(local_path)
    def save_checkpoint(self, local_path, global_step=0, max_ckpt_to_keep=None, **kwargs):
        os.makedirs(local_path)
        torch.save({"step": global_step}, os.path.join(local_path, "model.pt"))
        self.manager.register_checkpoint(local_path, max_ckpt_to_keep)
    def prepare_checkpoint_retention(self, loaded_path):
        method = getattr(self.manager, "record_loaded_checkpoint", None)
        if method is not None: method(loaded_path)
    def finalize_checkpoint_retention(self, new_path, max_ckpt_to_keep=None):
        method = getattr(self.manager, "finalize_loaded_checkpoint_retention", None)
        if method is not None: method(new_path, max_ckpt_to_keep)

with tempfile.TemporaryDirectory() as root:
    old = os.path.join(root, "global_step_1")
    os.makedirs(old)
    torch.save({"position": 1}, os.path.join(old, "data_0.pt"))
    open(os.path.join(root, "latest_checkpointed_iteration.txt"), "w").write("1")
    with patch("torch.distributed.get_rank", return_value=0), patch(
        "torch.distributed.get_world_size", return_value=1
    ):
        manager = BaseCheckpointManager(model=object(), optimizer=object())
    handler = CheckpointHandler(
        engine=Engine(manager), train_dataloader=Dataloader(), default_local_dir=root,
        max_ckpt_to_keep=1, mode=OrchestrationMode.RAY,
    )
    assert handler.load_checkpoint() == 1
    handler.save_checkpoint(2)
    print(sorted(name for name in os.listdir(root) if name.startswith("global_step_")))
```

The complete handoff reproducer supplies the small compatibility hooks needed to run unchanged against both revisions and repeats the lifecycle three times:

```bash
PYTHONPATH=/path/to/upstream-main python reproduce.py --expect bug
PYTHONPATH=/path/to/proposed-fix python reproduce.py --expect fixed
```

On main, all three runs produce:

```text
retained_after_restart=['global_step_1', 'global_step_2']
```

The same behavior occurs in a real restart:

1. Train Qwen3.5-0.8B for one SFT step with `trainer.max_ckpt_to_keep=1` and `trainer.resume_mode=disable`.
2. Terminate the process normally.
3. Start a new Ray/FSDP2 process with the same checkpoint directory, `trainer.resume_mode=auto`, and train step 2.
4. Inspect the tracker and `global_step_*` directories.

The GPU result below used the same lifecycle with a local Qwen3.5-0.8B checkpoint, four FSDP2 ranks, synchronous saves, and a fresh output directory for each A/B side.

## Expected behavior

After step 2 is completely saved, only `global_step_2` should remain because `max_ckpt_to_keep=1` requests one retained checkpoint.

## Actual behavior

The tracker advances to step 2, but both complete checkpoints remain:

```text
tracker=2
steps=[global_step_1, global_step_2]
logical_bytes=20,518,092,555
```

The four-GPU baseline contained all rank 0–3 model, optimizer, and extra-state shards for both steps. All 26 `.pt` files passed CRC validation, confirming that the extra directory was a complete old checkpoint rather than a partial write.

## Root cause

`BaseCheckpointManager` initializes `previous_saved_paths = []` in every new process. Retention in `ensure_checkpoint_capacity()` and `register_checkpoint()` only considers this in-memory list. The SFT resume path loads the tracker-selected checkpoint but never adds that successfully loaded path back to retention state, so the first save after every restart sees no previous checkpoint to rotate.

## Impact

- Affects SFT jobs using FSDP/FSDP2, `max_ckpt_to_keep=1`, and either `resume_mode=auto` or a `resume_path` that resolves into the same `default_local_dir` checkpoint series.
- Common triggers include preemption, node failure, planned maintenance, and manually resumed long-running training.
- The real 852.99M-parameter reproduction retained 20.518 GB instead of 10.259 GB after one restart. Each additional restart can leave another model plus optimizer checkpoint behind on main.
- For multi-billion-parameter models, one leaked checkpoint can be tens or hundreds of gigabytes depending on precision and optimizer layout.

This report intentionally does not claim that PPO actor/critic checkpoints, Megatron/TorchTitan, asynchronous saving, or `max_ckpt_to_keep > 1` are fixed by the proposed patch.

## Proposed fix

After a local FSDP checkpoint successfully loads, remember that exact path on rank 0. When SFT later completes a newer checkpoint in the same `default_local_dir` series, remove the loaded checkpoint only if all of the following hold:

- `max_ckpt_to_keep == 1`;
- the new FSDP checkpoint registered successfully;
- dataloader state and the atomic tracker update completed;
- any configured HDFS copy reported success;
- all SPMD ranks reached the existing checkpoint-save barrier;
- the new numeric step is greater than the loaded step.

Do not scan sibling directories. External resume paths, same/older steps, unknown directories, non-FSDP engines, and other retention values remain untouched.

## Validation of the proposed fix

The same four-GPU A/B produced:

| Result | main | proposed fix |
|---|---:|---:|
| Tracker | 2 | 2 |
| Retained steps | 1 and 2 | only 2 |
| Logical bytes | 20,518,092,555 | 10,259,046,150 |
| `.pt` files CRC-checked | 26 | 13 |

The fix saved 10,259,046,405 bytes (9.554 GiB, 50.0%) in this two-checkpoint reproduction. A second fresh process then loaded fixed `global_step_2` model, optimizer, RNG, and scheduler state, trained step 3, and left only `global_step_3`; all 13 resulting `.pt` files passed CRC validation. All five training phases exited with status 0, and no traceback, OOM, NCCL error, worker crash, NaN, or Inf appeared.

## Duplicate search

Open Issue and PR searches on 2026-08-13 for `max_ckpt_to_keep restart`, `max_ckpt_to_keep resume`, `previous_saved_paths`, `checkpoint retention restart`, and `old checkpoints after resume` found no report or implementation of this exact problem.

Related work is distinct:

- #4873 prevents deleting the current checkpoint before a replacement has finished saving in the same process; it does not restore retention state after a process restart.
- #724 introduced `max_ckpt_to_keep`.
- #2292 added SFT FSDP resume/retention support, but the loaded checkpoint is still not registered after restart.
- #7363 concerns Megatron SFT asynchronous checkpoint finalization, not FSDP retention history.
